{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1422a5f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-12-07\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "import os\n",
    "\n",
    "from datetime import date\n",
    "print(date.today())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba8dbb48",
   "metadata": {},
   "source": [
    "folder contents\n",
    "```bash\n",
    "dataset_nc2017kremer\n",
    "├── [9.4M]  S1_41467_2017_BFncomms15824_MOESM390_ESM.txt\n",
    "├── [ 30K]  S2_41467_2017_BFncomms15824_MOESM391_ESM.txt  \n",
    "├── [5.2K]  S3_41467_2017_BFncomms15824_MOESM392_ESM.txt   \n",
    "├── [204K]  S6_41467_2017_BFncomms15824_MOESM395_ESM.txt  \n",
    "├── [ 95K]  S7_41467_2017_BFncomms15824_MOESM396_ESM.txt  \n",
    "├── [4.1K]  S8_41467_2017_BFncomms15824_MOESM397_ESM.txt  \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0dad6cee",
   "metadata": {},
   "outputs": [],
   "source": [
    "workdir = '/mnt/disk7t/xwj/axolotl_rev'\n",
    "datasetdir = f'{workdir}/dataset_nc2017kremer'\n",
    "\n",
    "os.chdir(datasetdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b99043",
   "metadata": {},
   "source": [
    "### preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dca7a673",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27682, 119)\n",
      "(14409, 119)\n",
      "(11048, 119)\n"
     ]
    }
   ],
   "source": [
    "file = 'S1_41467_2017_BFncomms15824_MOESM390_ESM.txt' # RNA read count table. (TXT 9604 kb)\n",
    "cts = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "cts.index.name, cts.columns.name = 'Gene', 'Sample'\n",
    "# cts columns are RNA_ID, index are gene names.\n",
    "print(cts.shape)\n",
    "\n",
    "# whole sample set\n",
    "min_reads = 5\n",
    "cts14k = cts[ (cts > min_reads).any(axis=1) ].copy() # 14k genes\n",
    "print(cts14k.shape)\n",
    "\n",
    "# subset use non-zero genes only\n",
    "cts11k = cts.loc[ (cts > 0).all(axis=1) ].copy() # 11k genes\n",
    "cts11k.index = [ g.replace(' ','') for g in cts11k.index ]\n",
    "cts11k.index.name, cts11k.columns.name = 'Gene', 'Sample'\n",
    "cts11k.columns = [f'X{s}' if not s.startswith('MUC') else s for s in cts11k.columns]\n",
    "\n",
    "print(cts11k.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8729297b",
   "metadata": {},
   "source": [
    "## NMD genes + 6 valid genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6d69a92d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 LYRM7 MTO1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Gene</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FB49665</td>\n",
       "      <td>GTPBP3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;frame-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FB58027</td>\n",
       "      <td>NDUFB3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FB60592</td>\n",
       "      <td>MGME1</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FB61691</td>\n",
       "      <td>DNAJC3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FB61695</td>\n",
       "      <td>DNAJC3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FB61818</td>\n",
       "      <td>SFXN4</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;spl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FB68607</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FB73897</td>\n",
       "      <td>NBAS</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;fra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FB73898</td>\n",
       "      <td>IARS</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FB75110</td>\n",
       "      <td>NBAS</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FB76065</td>\n",
       "      <td>NSUN3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=coding;stop;RNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FB78659</td>\n",
       "      <td>PURA</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FB80304</td>\n",
       "      <td>SUCO</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;del...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FB80856</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FB80866</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FB80869</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FB81539</td>\n",
       "      <td>TXN2</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FB71215</td>\n",
       "      <td>LYRM7</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;sto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FB71215</td>\n",
       "      <td>MTO1</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=frame-shift;sto...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample    Gene                                            comment\n",
       "0   FB49665  GTPBP3  pred_effect=nmd;VARIANT_EFFECT=missense;frame-...\n",
       "1   FB58027  NDUFB3  pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...\n",
       "2   FB60592   MGME1  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...\n",
       "3   FB61691  DNAJC3  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...\n",
       "4   FB61695  DNAJC3  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...\n",
       "5   FB61818   SFXN4  pred_effect=nmd;VARIANT_EFFECT=frame-shift;spl...\n",
       "6   FB68607   CDKL5  pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...\n",
       "7   FB73897    NBAS  pred_effect=nmd;VARIANT_EFFECT=frame-shift;fra...\n",
       "8   FB73898    IARS  pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...\n",
       "9   FB75110    NBAS  pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...\n",
       "10  FB76065   NSUN3  pred_effect=nmd;VARIANT_EFFECT=coding;stop;RNA...\n",
       "11  FB78659    PURA  pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...\n",
       "12  FB80304    SUCO  pred_effect=nmd;VARIANT_EFFECT=frame-shift;del...\n",
       "13  FB80856    DPYD  pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...\n",
       "14  FB80866    DPYD  pred_effect=nmd;VARIANT_EFFECT=frame-shift;RNA...\n",
       "15  FB80869    DPYD  pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...\n",
       "16  FB81539    TXN2  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...\n",
       "17  FB71215   LYRM7  pred_effect=nmd;VARIANT_EFFECT=frame-shift;sto...\n",
       "18  FB71215    MTO1  pred_effect=nmd;VARIANT_EFFECT=frame-shift;sto..."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 这里有19个DNA提示NMD的Fibroblast样本,共20个基因（其中FB71215样本有两个基因)\n",
    "# RNAseq提示Aberrant Expression的样本不在其中\n",
    "file = 'S3_41467_2017_BFncomms15824_MOESM392_ESM.txt' # Sample annotation. (TXT 5 kb). \n",
    "# 'DIAGNOSIS_GENE'\n",
    "# 'EXOME_ENRICHMENT'\n",
    "# 'HAS_EXOME'\n",
    "# 'HAS_PROTEOME',\n",
    "# 'VARIANT_EFFECT' ( )\n",
    "# 'RNA_DEFECT_FOR_DIAGNOSIS_GENE' # 判定为诊断基因的那些基因名称 aberrant_splicing 8,  MAE  4, aberrant_expression  4\n",
    "# 'NUM_WES_SIGNI',\n",
    "# 'NUM_RNA_ABER_EXP_SIGNI'\n",
    "# 'NUM_RNA_ABER_SPLICE_SIGNI'\n",
    "# 'NUM_MAE_SIGNI'\n",
    "\n",
    "tab2 = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "\n",
    "key_effect = ['stop','frame-shift']\n",
    "tab2['pred_effect'] = None\n",
    "for ke in key_effect:\n",
    "    for idx, row in tab2.iterrows():\n",
    "        # 确保 VARIANT_EFFECT 是字符串类型\n",
    "        variant_effect = str(row['VARIANT_EFFECT'])\n",
    "        if ke in variant_effect:\n",
    "            tab2.loc[idx, 'pred_effect'] = 'nmd'\n",
    "\n",
    "\n",
    "# Part 1 of potential expression outliers\n",
    "cols = ['DIAGNOSIS_GENE', 'pred_effect','VARIANT_EFFECT',  'RNA_DEFECT_FOR_DIAGNOSIS_GENE']\n",
    "selected_cols = cols[1:]\n",
    "outlier_sample_col = 'Sample'\n",
    "outlier_gene_col = 'Gene'\n",
    "pred_nmd = tab2.query('pred_effect== \"nmd\"')[cols].copy()\n",
    "# for cols in ['VARIANT_EFFECT', 'pred_effect', 'RNA_DEFECT_FOR_DIAGNOSIS_GENE']:\n",
    "\n",
    "result = [\n",
    "    ';'.join([f\"{col}={row[col] if pd.notna(row[col]) else None }\" for col in selected_cols])\n",
    "    for _, row in pred_nmd.iterrows()\n",
    "]\n",
    "pred_nmd = pred_nmd.reset_index().rename(columns = {'DIAGNOSIS_GENE': outlier_gene_col, 'FIBROBLAST_ID': outlier_sample_col}).drop(columns=selected_cols)\n",
    "pred_nmd['comment'] = result\n",
    "pred_nmd[outlier_sample_col] = 'FB' + pred_nmd[outlier_sample_col].astype(str)\n",
    "\n",
    "# 一个病人有两个基因被判定为诊断基因，需要复制为two rows\n",
    "new_row = []\n",
    "del_row_idx = []\n",
    "for idx, row in pred_nmd.iterrows():\n",
    "    if \";\" in row['Gene']:\n",
    "        g1, g2 = row['Gene'].split(\";\")\n",
    "        # Sample\tGene\tcomment\n",
    "        new_row.append([row['Sample'], g1, row['comment']])\n",
    "        new_row.append([row['Sample'], g2, row['comment']])\n",
    "        del_row_idx.append(idx)\n",
    "        print(idx,g1, g2)\n",
    "        \n",
    "pred_nmd = pd.concat([pred_nmd.drop(del_row_idx), \n",
    "                      pd.DataFrame(new_row, columns = pred_nmd.columns)], \n",
    "                     ignore_index=True)\n",
    "pred_nmd = pred_nmd[pred_nmd['Gene'].isin(cts.index)].reset_index(drop=True)\n",
    "pred_nmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64a98448",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a558f40a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(119, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Gene</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FB58955</td>\n",
       "      <td>CLPP</td>\n",
       "      <td>WES_SIGNI=True;RNA_ABER_EXP_SIGNI=False;MAE_IS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FB35791</td>\n",
       "      <td>TIMMDC1</td>\n",
       "      <td>WES_SIGNI=None;RNA_ABER_EXP_SIGNI=True;MAE_IS_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FB66744</td>\n",
       "      <td>TIMMDC1</td>\n",
       "      <td>WES_SIGNI=None;RNA_ABER_EXP_SIGNI=True;MAE_IS_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FB73804</td>\n",
       "      <td>MGST1</td>\n",
       "      <td>WES_SIGNI=None;RNA_ABER_EXP_SIGNI=True;MAE_IS_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FB80256</td>\n",
       "      <td>ALDH18A1</td>\n",
       "      <td>WES_SIGNI=True;RNA_ABER_EXP_SIGNI=True;MAE_IS_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FB62346</td>\n",
       "      <td>MCOLN1</td>\n",
       "      <td>AGE slightly inferior of the thresholds</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample      Gene                                            comment\n",
       "0  FB58955      CLPP  WES_SIGNI=True;RNA_ABER_EXP_SIGNI=False;MAE_IS...\n",
       "1  FB35791   TIMMDC1  WES_SIGNI=None;RNA_ABER_EXP_SIGNI=True;MAE_IS_...\n",
       "2  FB66744   TIMMDC1  WES_SIGNI=None;RNA_ABER_EXP_SIGNI=True;MAE_IS_...\n",
       "3  FB73804     MGST1  WES_SIGNI=None;RNA_ABER_EXP_SIGNI=True;MAE_IS_...\n",
       "4  FB80256  ALDH18A1  WES_SIGNI=True;RNA_ABER_EXP_SIGNI=True;MAE_IS_...\n",
       "5  FB62346    MCOLN1            AGE slightly inferior of the thresholds"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# undiagnosed patients. Part 2 of potential expression outliers\n",
    "file = 'S7_41467_2017_BFncomms15824_MOESM396_ESM.txt' # Candidate genes for undiagnosed patients for all 3 strategies. (TXT 94 kb)\n",
    "# tab7 have fibroblast ids only \n",
    "tab7 = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "\n",
    "file = 'S8_41467_2017_BFncomms15824_MOESM397_ESM.txt' # RNA sample annotation for read count normalization. (TXT 4 kb)\n",
    "# used for sample id mapping. fibroblast id to RNA_ID\n",
    "tab8 = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "tab8['FB_ID'] = 'FB' + tab8.index.astype(str)\n",
    "tab8['RNA_ID'] = [f'X{s}' if not s.startswith('MUC') else s for s in tab8['RNA_ID']]\n",
    "tab8 = tab8.set_index('FB_ID')#.loc[71215,:]\n",
    "print(tab8.shape)\n",
    "# part 2 of potential expression outliers. DNA and Protein validation evidence.\n",
    "\n",
    "genes = [\"TIMMDC1\", \"MGST1\", \"CLPP\", \"ALDH18A1\",\"MCOLN1\"] \n",
    "# The above cases validated by Proteomics/western blot as described in Nature Communications (2017). also discussed  by OUTRIDER\n",
    "cols = ['HGNCID', \"WES_SIGNI\",'RNA_ABER_EXP_SIGNI', 'MAE_IS_SIGNI', 'RNA_ABER_SPLICE_SIGNI']\n",
    "selected_cols = cols[1:] \n",
    "outlier_sample_col = 'Sample'\n",
    "outlier_gene_col = 'Gene'\n",
    "\n",
    "val_protein = tab7.query('HGNCID in @genes')[cols].copy()\n",
    "\n",
    "result = [\n",
    "    ';'.join([f\"{col}={row[col] if pd.notna(row[col]) else None }\" for col in selected_cols])\n",
    "    for _, row in val_protein.iterrows()\n",
    "]\n",
    "val_protein = val_protein.reset_index().rename(columns = {'HGNCID': outlier_gene_col, 'FIBROBLAST_ID': outlier_sample_col}).drop(columns=selected_cols)\n",
    "val_protein['comment'] = result\n",
    "val_protein[outlier_sample_col] = 'FB' + val_protein[outlier_sample_col].astype(str)\n",
    "val_protein.loc[len(val_protein)] = [ tab8.query('RNA_ID == \"MUC1361\"').index.values[0], 'MCOLN1','AGE slightly inferior of the thresholds']\n",
    "val_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75b87321",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4b399e04",
   "metadata": {},
   "source": [
    "# Write counts file & outlier file & task config file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ca7f4119",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use 11k\n",
    "cts = cts11k.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "edd76634",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28, 7) (22, 7) (6, 7)\n"
     ]
    }
   ],
   "source": [
    "def func(raw): # 改成用RNA_ID n=119\n",
    "    df_outlier = tab8.loc[raw['Sample'],:].copy()\n",
    "    for idx, row in df_outlier.iterrows():\n",
    "        df_outlier.loc[idx, ['Gene','comment']]  = raw.set_index('Sample').loc[row.name].values\n",
    "    df_outlier = df_outlier.reset_index().rename(columns={'RNA_ID':'Sample'})\n",
    "    return df_outlier\n",
    "\n",
    "#  Part 1 of potential expression outliers\n",
    "# outlier_pred = pred_nmd\n",
    "outlier_pred = func(pred_nmd)\n",
    "outlier_val = func(val_protein)\n",
    "#  Part 2 of potential expression outliers. DNA and Protein validation evidence.\n",
    "outlier_pred_and_val = func(pd.concat([pred_nmd, val_protein], ignore_index=True))\n",
    "\n",
    "# df_outlier = tab8.loc[outlier_pred_and_val['Sample'],:].copy()\n",
    "# for idx, row in df_outlier.iterrows():\n",
    "#     df_outlier.loc[idx, ['Gene','comment']]  = outlier_pred_and_val.set_index('Sample').loc[row.name].values\n",
    "# df_outlier = df_outlier.reset_index().rename(columns={'RNA_ID':'Sample'})\n",
    "\n",
    "assert outlier_pred_and_val['Sample'].isin(cts.columns).all()\n",
    "assert outlier_pred['Sample'].isin(cts.columns).all()\n",
    "print(outlier_pred_and_val.shape, outlier_pred.shape, outlier_val.shape)\n",
    "# outlier_pred_and_val = df_outlier.copy()\n",
    "\n",
    "# ctsfile = f'{workdir}/gtex_processed/cts_{tissue_id}_s{cts.shape[1]}_g{cts.shape[0]}.tsv.gz'\n",
    "# outlierfile = f'{workdir}/gtex_processed/outlier_{tissue_id}_sg{outlier.shape[0]}.tsv'\n",
    "tissue_id = 'FB'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9176d445",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "id": "da0d0a2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/disk7t/xwj/axolotl_rev/dataset_nc2017kremer/outlier_FB_sg6.tsv\n"
     ]
    }
   ],
   "source": [
    "outlierfile3 =f'{datasetdir}/outlier_{tissue_id}_sg{outlier_val.shape[0]}.tsv'\n",
    "print(outlierfile3)\n",
    "outlier_val.to_csv(outlierfile3, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "id": "e40a1d1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FB_ID</th>\n",
       "      <th>Sample</th>\n",
       "      <th>SEX</th>\n",
       "      <th>RNA_HOX_GROUP</th>\n",
       "      <th>RNA_BATCH_GROUP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FB49665</td>\n",
       "      <td>X74172</td>\n",
       "      <td>male</td>\n",
       "      <td>magenta</td>\n",
       "      <td>orange</td>\n",
       "      <td>GTPBP3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;frame-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FB58027</td>\n",
       "      <td>X76630</td>\n",
       "      <td>female</td>\n",
       "      <td>magenta</td>\n",
       "      <td>purple</td>\n",
       "      <td>NDUFB3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FB60592</td>\n",
       "      <td>X76622</td>\n",
       "      <td>male</td>\n",
       "      <td>magenta</td>\n",
       "      <td>purple</td>\n",
       "      <td>MGME1</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FB61691</td>\n",
       "      <td>X76636</td>\n",
       "      <td>male</td>\n",
       "      <td>magenta</td>\n",
       "      <td>darkgreen</td>\n",
       "      <td>DNAJC3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FB61695</td>\n",
       "      <td>X61695</td>\n",
       "      <td>male</td>\n",
       "      <td>magenta</td>\n",
       "      <td>orange</td>\n",
       "      <td>DNAJC3</td>\n",
       "      <td>pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     FB_ID  Sample     SEX RNA_HOX_GROUP RNA_BATCH_GROUP    Gene  \\\n",
       "0  FB49665  X74172    male       magenta          orange  GTPBP3   \n",
       "1  FB58027  X76630  female       magenta          purple  NDUFB3   \n",
       "2  FB60592  X76622    male       magenta          purple   MGME1   \n",
       "3  FB61691  X76636    male       magenta       darkgreen  DNAJC3   \n",
       "4  FB61695  X61695    male       magenta          orange  DNAJC3   \n",
       "\n",
       "                                             comment  \n",
       "0  pred_effect=nmd;VARIANT_EFFECT=missense;frame-...  \n",
       "1  pred_effect=nmd;VARIANT_EFFECT=missense;stop;R...  \n",
       "2  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...  \n",
       "3  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...  \n",
       "4  pred_effect=nmd;VARIANT_EFFECT=stop;RNA_DEFECT...  "
      ]
     },
     "execution_count": 496,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Table output\n",
    "outlier_pred_and_val.to_excel(f'{workdir}/result/table/s2_kremer119_outlier28.xlsx', \n",
    "                              sheet_name='s2_kremer119_outlier28',index=True)\n",
    "outlier_pred_and_val.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f1da3866",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUC1350</th>\n",
       "      <th>MUC1344</th>\n",
       "      <th>MUC1365</th>\n",
       "      <th>MUC1396</th>\n",
       "      <th>MUC1404</th>\n",
       "      <th>MUC1361</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CLPP</th>\n",
       "      <td>581</td>\n",
       "      <td>980</td>\n",
       "      <td>1197</td>\n",
       "      <td>1337</td>\n",
       "      <td>991</td>\n",
       "      <td>940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMMDC1</th>\n",
       "      <td>2500</td>\n",
       "      <td>300</td>\n",
       "      <td>282</td>\n",
       "      <td>2403</td>\n",
       "      <td>2568</td>\n",
       "      <td>2303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMMDC1</th>\n",
       "      <td>2500</td>\n",
       "      <td>300</td>\n",
       "      <td>282</td>\n",
       "      <td>2403</td>\n",
       "      <td>2568</td>\n",
       "      <td>2303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGST1</th>\n",
       "      <td>4432</td>\n",
       "      <td>3955</td>\n",
       "      <td>3960</td>\n",
       "      <td>373</td>\n",
       "      <td>7887</td>\n",
       "      <td>6760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDH18A1</th>\n",
       "      <td>5664</td>\n",
       "      <td>6268</td>\n",
       "      <td>3704</td>\n",
       "      <td>5878</td>\n",
       "      <td>2642</td>\n",
       "      <td>4810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCOLN1</th>\n",
       "      <td>700</td>\n",
       "      <td>356</td>\n",
       "      <td>662</td>\n",
       "      <td>1061</td>\n",
       "      <td>1138</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          MUC1350  MUC1344  MUC1365  MUC1396  MUC1404  MUC1361\n",
       "Gene                                                          \n",
       "CLPP          581      980     1197     1337      991      940\n",
       "TIMMDC1      2500      300      282     2403     2568     2303\n",
       "TIMMDC1      2500      300      282     2403     2568     2303\n",
       "MGST1        4432     3955     3960      373     7887     6760\n",
       "ALDH18A1     5664     6268     3704     5878     2642     4810\n",
       "MCOLN1        700      356      662     1061     1138      150"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cts.loc[outlier_val['Gene'], outlier_val['Sample'] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2d97f4ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t00_FB_s119_g14409 (14409, 119)\n",
      "Fibroblast/FB/min_reads=5 stable_ngene=14409 nsample=119\n",
      "t00_FB_s119_g11048 (11048, 119)\n",
      "Fibroblast/FB/min_reads=5 stable_ngene=11048 nsample=119\n"
     ]
    }
   ],
   "source": [
    "for use in ['t00_FB_s119_g14409', 't00_FB_s119_g11048']:\n",
    "    if use == 't00_FB_s119_g14409':\n",
    "        cts = cts14k.copy()\n",
    "    elif use == 't00_FB_s119_g11048': # use = 't00_FB_s119_g11048' # all genes have minreads 1, subset random set from this cts table\n",
    "        cts = cts11k.copy()\n",
    "    print(use, cts.shape)\n",
    "    tissue_name_mapping = pd.DataFrame({\n",
    "        'TISSUE_ID': ['FB'],\n",
    "        'TISSUE_NAME': ['Fibroblast'],\n",
    "    })\n",
    "\n",
    "    tissues = pd.DataFrame(\n",
    "        index = tissue_name_mapping.index,\n",
    "        columns = ['TISSUE_ID','TISSUE_NAME','N_SAMPLE','N_GENE','CTS_FILE', 'OUTLIER_FILE',])\n",
    "\n",
    "    # 各组织cts表达矩阵和outlier\n",
    "    for idx, row in tissue_name_mapping.iterrows():\n",
    "        tissue_id, tissue_name = row['TISSUE_ID'],  row['TISSUE_NAME']\n",
    "        \n",
    "        # (1) cts: expression matrix\n",
    "        # (2) outlier: true outlier gene-sample pair in cts.\n",
    "        ctsfile = f'{datasetdir}/cts_{tissue_id}_s{cts.shape[1]}_g{cts.shape[0]}.tsv.gz'\n",
    "        cts.to_csv(ctsfile, sep='\\t')\n",
    "        # outliers: part1 + part2\n",
    "        outlierfile =f'{datasetdir}/outlier_{tissue_id}_sg{outlier_pred_and_val.shape[0]}.tsv'\n",
    "        outlier_pred_and_val.to_csv(outlierfile, sep='\\t')\n",
    "        # outliers: part1 only\n",
    "        outlierfile2 =f'{datasetdir}/outlier_{tissue_id}_sg{outlier_pred.shape[0]}.tsv'\n",
    "        outlier_pred.to_csv(outlierfile2, sep='\\t')\n",
    "        \n",
    "        print(f'{tissue_name}/{tissue_id}/min_reads={min_reads} stable_ngene={cts.shape[0]} nsample={cts.shape[1]}')\n",
    "\n",
    "        # print(outlierfile,ctsfile)\n",
    "        tissues.loc[idx, :] = [ tissue_id, tissue_name, cts.shape[1], cts.shape[0],ctsfile, outlierfile]\n",
    "        \n",
    "    tissues.to_csv(f'tissues_{use}.tsv', sep='\\t')\n",
    "    tissues.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "acda369f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TISSUE_ID</th>\n",
       "      <th>TISSUE_NAME</th>\n",
       "      <th>N_SAMPLE</th>\n",
       "      <th>N_GENE</th>\n",
       "      <th>CTS_FILE</th>\n",
       "      <th>OUTLIER_FILE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FB</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>119</td>\n",
       "      <td>14409</td>\n",
       "      <td>/mnt/disk7t/xwj/axolotl_rev/dataset_nc2017krem...</td>\n",
       "      <td>/mnt/disk7t/xwj/axolotl_rev/dataset_nc2017krem...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  TISSUE_ID TISSUE_NAME  N_SAMPLE  N_GENE  \\\n",
       "0        FB  Fibroblast       119   14409   \n",
       "\n",
       "                                            CTS_FILE  \\\n",
       "0  /mnt/disk7t/xwj/axolotl_rev/dataset_nc2017krem...   \n",
       "\n",
       "                                        OUTLIER_FILE  \n",
       "0  /mnt/disk7t/xwj/axolotl_rev/dataset_nc2017krem...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "use = 't00_FB_s119_g14409' # all genes have minreads 1, subset random set from here\n",
    "\n",
    "tissues = pd.read_csv(f'tissues_{use}.tsv', sep='\\t', index_col=0)\n",
    "tissues "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "ed4a849b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# level 1\n",
    "output_path = f'{workdir}/result/dataset_nc2017kremer'\n",
    "# level 2\n",
    "samples_path = f'{output_path}/samples'\n",
    "task_config_path = f'{output_path}/task_config'\n",
    "task_output_path = f'{output_path}/task_output'\n",
    "metric_output_path = f'{output_path}/metric'\n",
    "\n",
    "# print(workdir, samples_path, task_config_path, task_output_path, metric_output_path)\n",
    "os.system(f'mkdir -p {samples_path} {task_config_path} {task_output_path} {metric_output_path}')\n",
    "# os.system(f'chmod --silent -R 777 {task_output_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4317346a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare task config file.\n",
    "# task_config table have columns indicating output filenames of different methods\n",
    "\n",
    "for i, row in tissues.iterrows():\n",
    "    \n",
    "    t  = row.TISSUE_ID\n",
    "    tn = row.TISSUE_NAME\n",
    "    ns = row.N_SAMPLE\n",
    "    ng = row.N_GENE\n",
    "    ctsfile = row.CTS_FILE\n",
    "    outlierfile = row.OUTLIER_FILE\n",
    "\n",
    "    prefix = f't{i:02d}_{t}_s{ns}_g{ng}' # id, number of samples, number of genes.\n",
    "    \n",
    "    cols = ['Dname','cts','samples','MyMethod','OUTRIDER','ABEILLE','OUTSINGLE']\n",
    "    tasks = [0] # list of parallel tasks\n",
    "    task_config = pd.DataFrame(index=tasks, columns=cols)\n",
    "    task_config.index.name = 'task'\n",
    "\n",
    "    task = 0 # 默认是全部样本, 所以only one task for this tissue  \n",
    "    task_config.loc[task, 'Dname' ] = t\n",
    "    task_config.loc[task, 'cts' ] = ctsfile\n",
    "    task_config.loc[task, 'samples' ] = f'{samples_path}/{prefix}.txt'\n",
    "    # create filenames\n",
    "    task_config.loc[task, 'MyMethod'] = f'{task_output_path}/{prefix}/{task:03d}_mymethod.txt.gz'\n",
    "    task_config.loc[task, 'OUTRIDER'] = f'{task_output_path}/{prefix}/{task:03d}_outrider.txt.gz'\n",
    "    task_config.loc[task, 'OUTSINGLE'] = f'{task_output_path}/{prefix}/{task:03d}_outsingle.txt.gz'\n",
    "    task_config.loc[task, 'ABEILLE'] = f'{task_output_path}/{prefix}/{task:03d}_abeille.txt.gz'\n",
    "\n",
    "    # 0. create config & output folder of parallel tasks\n",
    "    task_config.to_csv(f'{task_config_path}/{prefix}.config',sep='\\t')\n",
    "    os.system(f'mkdir -p {task_output_path}/{prefix}')\n",
    "    \n",
    "    # 1. sample ids of parallel tasks\n",
    "    # 将task specific样本列表作为one row添加到DataFrame中. 默认task是全部样本\n",
    "    cts = pd.read_csv(ctsfile, sep='\\t',index_col=0)\n",
    "    all_samples_df = pd.DataFrame(data=cts.columns.T.tolist()).transpose()\n",
    "    all_samples_df.index = task_config.index\n",
    "    all_samples_df.to_csv(f'{samples_path}/{prefix}.txt',sep='\\t')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd395a83",
   "metadata": {},
   "source": [
    "## subsampling to small sizes (min_reads = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "12a5448c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((21, 7), (6, 7))"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# positive samples\n",
    "part1_6 = outlier_val.copy()\n",
    "part2_18 = outlier_pred.drop_duplicates('Sample').copy()\n",
    "part2_18.shape, part1_6.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4d89cd6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tissues_t00_FB_s119_g11048.tsv'"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "use = 't00_FB_s119_g11048'\n",
    "tissues = pd.read_csv(f'tissues_{use}.tsv', sep='\\t', index_col=0)\n",
    "f'tissues_{use}.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "94ffce9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/disk7t/xwj/axolotl_rev/dataset_nc2017kremer'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "88083fbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 0 (6, 7) (21, 7) 73 100\n",
      "100 1 (6, 7) (21, 7) 73 100\n",
      "100 2 (6, 7) (21, 7) 73 100\n",
      "100 3 (6, 7) (21, 7) 73 100\n",
      "100 4 (6, 7) (21, 7) 73 100\n",
      "100 5 (6, 7) (21, 7) 73 100\n",
      "100 6 (6, 7) (21, 7) 73 100\n",
      "100 7 (6, 7) (21, 7) 73 100\n",
      "100 8 (6, 7) (21, 7) 73 100\n",
      "100 9 (6, 7) (21, 7) 73 100\n",
      "60 0 (6, 7) (21, 7) 33 60\n",
      "60 1 (6, 7) (21, 7) 33 60\n",
      "60 2 (6, 7) (21, 7) 33 60\n",
      "60 3 (6, 7) (21, 7) 33 60\n",
      "60 4 (6, 7) (21, 7) 33 60\n",
      "60 5 (6, 7) (21, 7) 33 60\n",
      "60 6 (6, 7) (21, 7) 33 60\n",
      "60 7 (6, 7) (21, 7) 33 60\n",
      "60 8 (6, 7) (21, 7) 33 60\n",
      "60 9 (6, 7) (21, 7) 33 60\n",
      "30 0 (6, 7) (21, 7) 3 30\n",
      "30 1 (6, 7) (21, 7) 3 30\n",
      "30 2 (6, 7) (21, 7) 3 30\n",
      "30 3 (6, 7) (21, 7) 3 30\n",
      "30 4 (6, 7) (21, 7) 3 30\n",
      "30 5 (6, 7) (21, 7) 3 30\n",
      "30 6 (6, 7) (21, 7) 3 30\n",
      "30 7 (6, 7) (21, 7) 3 30\n",
      "30 8 (6, 7) (21, 7) 3 30\n",
      "30 9 (6, 7) (21, 7) 3 30\n",
      "10 0 (6, 7) (4, 7) 0 10\n",
      "10 1 (6, 7) (4, 7) 0 10\n",
      "10 2 (6, 7) (4, 7) 0 10\n",
      "10 3 (6, 7) (4, 7) 0 10\n",
      "10 4 (6, 7) (4, 7) 0 10\n",
      "10 5 (6, 7) (4, 7) 0 10\n",
      "10 6 (6, 7) (4, 7) 0 10\n",
      "10 7 (6, 7) (4, 7) 0 10\n",
      "10 8 (6, 7) (4, 7) 0 10\n",
      "10 9 (6, 7) (4, 7) 0 10\n",
      "sampled_configs已保存到: /mnt/disk7t/xwj/axolotl_rev/result/dataset_nc2017kremer/task_config/t01_FB_s119_g11048_nmd_sampled_configs_dict.pkl\n"
     ]
    }
   ],
   "source": [
    "# 定义样本量列表和随机种子\n",
    "sample_sizes = [100, 60, 30, 10]\n",
    "n_replicates = 10  # 每个样本量重复10次\n",
    "random_seeds = range(n_replicates)  # 使用0-9作为随机种子\n",
    "\n",
    "# 从tissues中获取信息\n",
    "tissue_data = tissues.iloc[0]\n",
    "t = tissue_data[0]  # TISSUE_ID\n",
    "tn = tissue_data[1]  # TISSUE_NAME\n",
    "ns = tissue_data[2]  # N_SAMPLE\n",
    "ng = tissue_data[3]  # N_GENE\n",
    "ctsfile = tissue_data[4]  # CTS_FILE\n",
    "outlierfile = tissue_data[5]  # OUTLIER_FILE\n",
    "\n",
    "# 创建基础前缀\n",
    "# base_prefix = f't00_{t}_s{ns}_g{ng}' # Version1.  six valid must in sub samples \n",
    "base_prefix = f't01_{t}_s{ns}_g{ng}_nmd' # Version2.  six valid + equal number of NMD, must in sub samples\n",
    "\n",
    "cts = pd.read_csv(ctsfile, sep='\\t', index_col=0)\n",
    "all_samples = cts.columns.tolist()\n",
    "\n",
    "# positive samples\n",
    "part1_sampled = outlier_val.copy()\n",
    "part2_18 = outlier_pred.drop_duplicates('Sample').copy()\n",
    "\n",
    "# all negative samples\n",
    "negative_samples = list(set(all_samples) - set(part1_sampled['Sample'].tolist()) - set(part2_18['Sample'].tolist()) )\n",
    "\n",
    "# 创建字典存储所有抽样结果\n",
    "sampled_configs = {}\n",
    "\n",
    "# 对每个样本量进行抽样\n",
    "for size in sample_sizes:\n",
    "    # 存储当前样本量的所有抽样结果\n",
    "    size_configs = {}\n",
    "    \n",
    "    for seed in random_seeds:\n",
    "        if size <= 10:\n",
    "            part2_sampled = part2_18.sample(n = size - len(part1_sampled), random_state=seed)\n",
    "        else:\n",
    "            part2_sampled = part2_18.copy()\n",
    "            \n",
    "        n_negative = size - len(part1_sampled) - len(part2_sampled)\n",
    "        if n_negative: # 从阴性样本中随机抽样（数量 = 总样本量 - 阳性样本数量\n",
    "            # negative_selected =  np.random.choice(negative_samples, size=n_negative, replace=False)\n",
    "            negative_selected = pd.Series(negative_samples).sample(n = n_negative, random_state=seed).tolist() #, size=n_negative, replace=False)\n",
    "        else:\n",
    "            negative_selected = []\n",
    "        # 合并阳性样本和随机选择的阴性样本\n",
    "        selected_samples =  part1_sampled['Sample'].tolist() + part2_sampled['Sample'].tolist() + negative_selected\n",
    "        assert len(selected_samples), len(set(selected_samples))\n",
    "        print(size, seed, part1_sampled.shape, part2_sampled.shape, len(negative_selected), len(selected_samples))\n",
    "        \n",
    "        # 创建前缀\n",
    "        prefix = f'{base_prefix}_size{size}_seed{seed}'\n",
    "        \n",
    "        # 创建输出目录\n",
    "        output_dir = os.path.join(task_output_path, prefix)\n",
    "        os.makedirs(output_dir, exist_ok=True)\n",
    "        \n",
    "        # 创建样本列表文件\n",
    "        samples_file = os.path.join(samples_path, f'{prefix}.txt')\n",
    "        samples = pd.DataFrame(selected_samples).transpose()\n",
    "        samples.index.name = 'task'\n",
    "        samples.to_csv(samples_file, sep='\\t', header=True)\n",
    "        \n",
    "        # 创建config文件\n",
    "        cols = ['Dname', 'cts', 'samples', 'MyMethod', 'OUTRIDER', 'ABEILLE', 'OUTSINGLE']\n",
    "        task_config = pd.DataFrame(columns=cols)\n",
    "        task_config.index.name = 'task'\n",
    "        # 填充config数据\n",
    "        task = 0\n",
    "        task_config.loc[0, 'Dname'] = t\n",
    "        task_config.loc[0, 'cts'] = ctsfile\n",
    "        task_config.loc[0, 'samples'] = samples_file\n",
    "        task_config.loc[0, 'MyMethod'] = os.path.join(output_dir, f'{task:03d}_mymethod.txt.gz')\n",
    "        task_config.loc[0, 'OUTRIDER'] = os.path.join(output_dir, f'{task:03d}_outrider.txt.gz')\n",
    "        task_config.loc[0, 'OUTSINGLE'] = os.path.join(output_dir, f'{task:03d}_outsingle.txt.gz')\n",
    "        task_config.loc[0, 'ABEILLE'] = os.path.join(output_dir, f'{task:03d}_abeille.txt.gz')\n",
    "        # {prefix}/{task:03d}\n",
    "        \n",
    "        # 保存config文件\n",
    "        config_file = os.path.join(task_config_path, f'{prefix}.config')\n",
    "        task_config.to_csv(config_file, sep='\\t', index=False)\n",
    "        \n",
    "        # 存储结果\n",
    "        size_configs[f'seed_{seed}'] = {\n",
    "            'config': task_config,\n",
    "            'selected_samples': selected_samples,\n",
    "            'output_dir': output_dir\n",
    "        }\n",
    "    \n",
    "    # 将当前样本量的所有结果存入总字典\n",
    "    sampled_configs[f'size_{size}'] = size_configs\n",
    "\n",
    "# sampled_configs现在包含了所有抽样结果和对应的配置文件\n",
    "# 结构为：{sample_size: {seed: {'config': df, 'selected_samples': list, 'output_dir': str}}, ...}\n",
    "output_file = os.path.join(task_config_path, f'{base_prefix}_sampled_configs_dict.pkl')\n",
    "# 保存sampled_configs字典\n",
    "with open(output_file, 'wb') as f:\n",
    "    pickle.dump(sampled_configs, f)\n",
    "\n",
    "    print(f\"sampled_configs已保存到: {output_file}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99f8105b",
   "metadata": {},
   "source": [
    "### pct change from 0.08~0.24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "2a1bb752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t00_FB_s119_g11048_pct0.08\n",
      "8 92\n",
      "4 46\n",
      "sampled_configs已保存到: /mnt/disk7t/xwj/axolotl_rev/result/dataset_nc2017kremer/task_config/t00_FB_s119_g11048_pct0.08_sampled_configs_dict.pkl\n",
      "t00_FB_s119_g11048_pct0.16\n",
      "16 84\n",
      "8 42\n",
      "sampled_configs已保存到: /mnt/disk7t/xwj/axolotl_rev/result/dataset_nc2017kremer/task_config/t00_FB_s119_g11048_pct0.16_sampled_configs_dict.pkl\n",
      "t00_FB_s119_g11048_pct0.24\n",
      "24 76\n",
      "12 38\n",
      "sampled_configs已保存到: /mnt/disk7t/xwj/axolotl_rev/result/dataset_nc2017kremer/task_config/t00_FB_s119_g11048_pct0.24_sampled_configs_dict.pkl\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# 定义样本量列表和随机种子\n",
    "list_outlier_pct = [ 0.08, 0.16, 0.24,]\n",
    "sample_sizes = [100, 50,]\n",
    "n_replicates = 10  # 每个样本量重复10次\n",
    "random_seeds = range(n_replicates)  # 使用0-9作为随机种子\n",
    "\n",
    "# 从tissues中获取信息\n",
    "tissue_data = tissues.iloc[0]\n",
    "t = tissue_data[0]  # TISSUE_ID\n",
    "tn = tissue_data[1]  # TISSUE_NAME\n",
    "ns = tissue_data[2]  # N_SAMPLE\n",
    "ng = tissue_data[3]  # N_GENE\n",
    "ctsfile = tissue_data[4]  # CTS_FILE\n",
    "outlierfile = tissue_data[5]  # OUTLIER_FILE\n",
    "\n",
    "# 创建基础前缀\n",
    "# base_prefix = f't00_{t}_s{ns}_g{ng}' # Version1.  six valid must in sub samples \n",
    "base_prefix = f't01_{t}_s{ns}_g{ng}_nmd' # Version2.  six valid + equal number of NMD, must in sub samples\n",
    "\n",
    "cts = pd.read_csv(ctsfile, sep='\\t', index_col=0)\n",
    "all_samples = cts.columns.tolist()\n",
    "\n",
    "# positive samples\n",
    "part1_sampled = outlier_val.copy()\n",
    "part2_18 = outlier_pred.drop_duplicates('Sample').copy()\n",
    "\n",
    "# all negative samples\n",
    "negative_samples = list(set(all_samples) - set(part1_sampled['Sample'].tolist()) - set(part2_18['Sample'].tolist()) )\n",
    "i = 0\n",
    "for pct in list_outlier_pct: # 增加pct的变化\n",
    "    # 创建基础前缀\n",
    "    base_prefix = f't{i:02d}_{t}_s{ns}_g{ng}_pct{pct:.2f}'\n",
    "    print(base_prefix)\n",
    "    # 创建字典存储所有抽样结果\n",
    "    sampled_configs = {}\n",
    "\n",
    "    # 对每个样本量进行抽样\n",
    "    for size in sample_sizes:\n",
    "        # 存储当前样本量的所有抽样结果\n",
    "        size_configs = {}\n",
    "        n_positive = int(size * pct) \n",
    "        n_negative = size - n_positive\n",
    "        print(n_positive, n_negative)\n",
    "        \n",
    "        for seed in random_seeds:\n",
    "            negative_selected = pd.Series(negative_samples).sample(n = n_negative, random_state=seed*1000+n_negative).tolist() \n",
    "            # include all part1_sampled. part1+part = n_positive\n",
    "            if n_positive <= len(part1_sampled):\n",
    "                part1_sampled_enough = part1_sampled.sample(n = n_positive, random_state=seed*1000+n_positive)\n",
    "                part2_sampled = part2_18.sample(0)\n",
    "                selected_samples =  part1_sampled_enough['Sample'].tolist() + negative_selected\n",
    "            else:\n",
    "                part2_sampled = part2_18.sample(n = n_positive - len(part1_sampled), random_state=seed*1000+n_positive)            \n",
    "                # 合并阳性样本和随机选择的阴性样本\n",
    "                selected_samples =  part1_sampled['Sample'].tolist() + part2_sampled['Sample'].tolist() + negative_selected\n",
    "            assert len(selected_samples), len(set(selected_samples))\n",
    "            # print(size, seed, len(part2_sampled), len(negative_selected), len(selected_samples))\n",
    "            \n",
    "            # 创建前缀\n",
    "            prefix = f'{base_prefix}_size{size}_seed{seed}'\n",
    "            \n",
    "            # 创建输出目录\n",
    "            output_dir = os.path.join(task_output_path, prefix)\n",
    "            os.makedirs(output_dir, exist_ok=True)\n",
    "            \n",
    "            # 创建样本列表文件\n",
    "            samples_file = os.path.join(samples_path, f'{prefix}.txt')\n",
    "            samples = pd.DataFrame(selected_samples).transpose()\n",
    "            samples.index.name = 'task'\n",
    "            samples.to_csv(samples_file, sep='\\t', header=True)\n",
    "            \n",
    "            # 创建config文件\n",
    "            cols = ['Dname', 'cts', 'samples', 'MyMethod', 'OUTRIDER', 'ABEILLE', 'OUTSINGLE']\n",
    "            task_config = pd.DataFrame(columns=cols)\n",
    "            task_config.index.name = 'task'\n",
    "            # 填充config数据\n",
    "            task = 0\n",
    "            task_config.loc[0, 'Dname'] = t\n",
    "            task_config.loc[0, 'cts'] = ctsfile\n",
    "            task_config.loc[0, 'samples'] = samples_file\n",
    "            task_config.loc[0, 'MyMethod'] = os.path.join(output_dir, f'{task:03d}_mymethod.txt.gz')\n",
    "            task_config.loc[0, 'OUTRIDER'] = os.path.join(output_dir, f'{task:03d}_outrider.txt.gz')\n",
    "            task_config.loc[0, 'OUTSINGLE'] = os.path.join(output_dir, f'{task:03d}_outsingle.txt.gz')\n",
    "            task_config.loc[0, 'ABEILLE'] = os.path.join(output_dir, f'{task:03d}_abeille.txt.gz')\n",
    "            # {prefix}/{task:03d}\n",
    "            \n",
    "            # 保存config文件\n",
    "            config_file = os.path.join(task_config_path, f'{prefix}.config')\n",
    "            task_config.to_csv(config_file, sep='\\t', index=False)\n",
    "            \n",
    "            # 存储结果\n",
    "            size_configs[f'seed_{seed}'] = {\n",
    "                'config': task_config,\n",
    "                'selected_samples': selected_samples,\n",
    "                'output_dir': output_dir\n",
    "            }\n",
    "        \n",
    "        # 将当前样本量的所有结果存入总字典\n",
    "        sampled_configs[f'size_{size}'] = size_configs\n",
    "\n",
    "    # sampled_configs现在包含了所有抽样结果和对应的配置文件\n",
    "    # 结构为：{sample_size: {seed: {'config': df, 'selected_samples': list, 'output_dir': str}}, ...}\n",
    "    output_file = os.path.join(task_config_path, f'{base_prefix}_sampled_configs_dict.pkl')\n",
    "    # 保存sampled_configs字典\n",
    "    with open(output_file, 'wb') as f:\n",
    "        pickle.dump(sampled_configs, f)\n",
    "        print(f\"sampled_configs已保存到: {output_file}\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c00deb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3.12]",
   "language": "python",
   "name": "conda-env-py3.12-py"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
